### DISCUSSION TITLE
how to upx exclude dll/pyd files in sub-folder?

### DISCUSSION THREAD

**Discussion by mo-han** [pyinstaller/pyinstaller#6128] [Help]
my program depends on pyside2, after using pyinstaller, there is some Qt related .pyd files under the sub-folder `dist/myapp/PySide2`.
i need to exclude these .pyd files to workaround `ImportError: DLL load failed: Invalid access to memory location` issue, but `--upx-exclude` or `upx_exclude` have no effect when those .pyd file names are given. i guess it's the sub-folder that stops the procedure, but i have no idea how to fix this.
any infomation will be appreciated.

**Comment by rokm** [comment]
Looks like when considering the `upx_exclude` list, only file's basename is compared against the provided patterns:
https://github.com/pyinstaller/pyinstaller/blob/40f75e690cdcb30765f03639961e07736ab4b1ef/PyInstaller/building/utils.py#L183

So providing just the basenames should work:

`pyinstaller  --upx-exclude qtcore.pyd --upx-exclude qtgui.pyd --upx-exclude qtwidgets.pyd ...`

**Reply by mo-han** [reply]
Already tried basename, no good unfortunately (confused as well)

**Reply by rokm** [reply]
If you run with `--log-level DEBUG`, do you see these three files still processed in spite of adding `--upx-exclude qtcore.pyd --upx-exclude qtgui.pyd --upx-exclude qtwidgets.pyd`? (Do you perhaps need to add a `--clean` to force full rebuild?)

**Reply by mo-han** [reply]
@rokm 
OK, it's not the sub-folder, but the case sensitivity, that cause this. Before checking file names with `upx_exclude`, they are handled by `normcased`, which convert into lower case (on Windows OS). However those items in `upx_exclude` are not processed in the same way, and I provide them in the real form like `PyCore.pyd`.
This "feature" is not documented. Besides, if we use `normcase` on those basenames, then should also do the same thing to those in `upx_exclude`.

**Reply by rokm** [reply]
> This "feature" is not documented. Besides, if we use `normcase` on those basenames, then should also do the same thing to those in `upx_exclude`.

Indeed, `normcase` should be used consistently. Will take care of this "feature"...

**Reply by rokm** [reply]
The UPX exclude pattern matching has been revised in #6161. It now uses `pathlib.PurePath.match()` to match the file's source path against the given pattern(s). The comparison is therefore done with OS-default case sensitivity, it allows wildcard (`*`), and allows patterns to contain parent directories.  For example, you could now use `--upx-exclude "Qt*.dll"` and `--upx-exclude "PySide2\*.pyd"`.

The change was merged to `develop` and is scheduled to appear in the new PyInstaller 5.x series.
